libwallet/assets/dcr/agenda.go

```diff
@@ -0,0 +1,249 @@
+	ctx        context.Context
+	cancelSync context.CancelFunc
+
+	syncCallbacksMtx *sync.RWMutex // Pointer required to avoid copying literal values.
+	syncCallbacks    map[string]consensusAgendaSyncCallback
+}
+
+func NewConsensusAgenda(chainParams *chaincfg.Params, db *storm.DB) *ConsensusAgenda {
+	return &ConsensusAgenda{
+		chainParams:      chainParams,
+		db:               db,
+		mu:               &sync.RWMutex{},
+		syncCallbacksMtx: &sync.RWMutex{},
+		syncCallbacks:    make(map[string]consensusAgendaSyncCallback),
+	}
+}
+
+func (c *ConsensusAgenda) saveLastSyncedTimestamp(lastSyncedTimestamp int64) {
+	err := c.db.Set(configDBBkt, LastSyncedTimestampConfigKey, &lastSyncedTimestamp)
+	if err != nil {
+		log.Errorf("error setting config value for key: %s, error: %v", LastSyncedTimestampConfigKey, err)
+	}
+}
+
+func (c *ConsensusAgenda) GetLastSyncedTimestamp() (lastSyncedTimestamp int64) {
+	err := c.db.Get(configDBBkt, LastSyncedTimestampConfigKey, &lastSyncedTimestamp)
+	if err != nil && err != storm.ErrNotFound {
+		log.Errorf("error reading config value for key: %s, error: %v", LastSyncedTimestampConfigKey, err)
+	}
+	return lastSyncedTimestamp
+}
+
+func (c *ConsensusAgenda) saveOrOverwiteProposal(dataAgenda *DcrdataAgenda) error {
+	var oldAgenda DcrdataAgenda
+	err := c.db.One("Name", dataAgenda.Name, &oldAgenda)
+	if err != nil && err != storm.ErrNotFound {
+		return errors.Errorf("error checking if proposal was already indexed: %s", err.Error())
+	}
+
+	if oldAgenda.Name != "" {
+		// delete old record before saving new (if it exists)
+		_ = c.db.DeleteStruct(oldAgenda)
+	}
+
+	return c.db.Save(dataAgenda)
+}
+
+// GetAgendaRaw fetches and returns a agenda from the db
+func (c *ConsensusAgenda) getAgendaRaw(offset, limit int32, newestFirst bool) ([]DcrdataAgenda, error) {
+	query := c.db.Select()
+
+	if offset > 0 {
+		query = query.Skip(int(offset))
+	}
+
+	if limit > 0 {
+		query = query.Limit(int(limit))
+	}
+
+	if newestFirst {
+		query = query.OrderBy("StartTime").Reverse()
+	} else {
+		query = query.OrderBy("StartTime")
+	}
+
+	var agendas []DcrdataAgenda
+	err := query.Find(&agendas)
+	if err != nil && err != storm.ErrNotFound {
+		return nil, fmt.Errorf("error fetching proposals: %s", err.Error())
+	}
+
+	return agendas, nil
+}
+
+func (c *ConsensusAgenda) getAgendaDataFromURL() ([]*DcrdataAgenda, error) {
+	// Fetch high level agenda detail form dcrdata api.
+	var dcrdataAgenda []*DcrdataAgenda
+	host := dcrdataAgendasAPIMainnetURL
+	if c.chainParams.Net == wire.TestNet3 {
+		host = dcrdataAgendasAPITestnetURL
+	}
+
+	req := &utils.ReqConfig{
+		Method:  http.MethodGet,
+		HTTPURL: host,
+	}
+
+	if _, err := utils.HTTPRequest(req, &dcrdataAgenda); err != nil {
+		return nil, err
+	}
+	return dcrdataAgenda, nil
+}
+
+func (c *ConsensusAgenda) IsSyncing() bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.cancelSync != nil
+}
+
+func (c *ConsensusAgenda) StopSync() {
+	c.mu.RLock()
```

Comment:
Since you are potentially modifying the cancelSync field, it would be more appropriate to use a write lock (`c.mu.Lock()`) to ensure safe concurrent access